### ISSUE TITLE
Rails 7 - slow store time Carrierwave S3 AWS fog upload ONLY in localhost

### ISSUE THREAD

**Issue by robertoboldi** [carrierwaveuploader/carrierwave#2605]
I have switched to Rails 7 and I found some issues when I upload image in my localhost. 
It is mandatory to manage, import, rework and upload images on a local environment, and Images (and attachments) needs to be uploaded from localhost.
The system idles for 10/30 sec even for 5kb images, picture will be correctly loaded. 

Furthermore the server is also very slow when deleting images.

I underline that this happens only in localhost, not production.

My system is a last generation M1 MacBook pro.

Here is my code:
carrierwave.rb

    CarrierWave.configure do |config|

    config.fog_provider = "fog/aws"
    config.fog_credentials = {
        :provider               => 'AWS',                        # required
        :aws_access_key_id      => ENV['S3_ACCESS_KEY'],                        
        :aws_secret_access_key  => ENV['S3_SECRET_ACCESS_KEY'],    
        :region                 => ENV['S3_REGION']                          # optional, defaults to 'us-east-1' defaults to 'us-east-1'
    }
    config.fog_directory  = ENV['S3_BUCKET']                          # required
    # config.fog_public     = true                                   # optional, defaults to true
    config.fog_attributes = {'Cache-Control'=>'max-age=315576000'}  # optional, defaults to {}
    config.storage = :fog

image_uploader.rb

    class ImageUploader < CarrierWave::Uploader::Base
    include CarrierWave::MiniMagick
    storage :fog
    def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
    end
    version :thumb do
      process resize_to_fit: [300, 300]
    end
    version :default do 
      process resize_to_fit: [800,800]
    end
    version :big do 
      process resize_to_limit: [1200, nil]
    end
    end
    end

the issue is shown during uploads in my active admin page, with standard form uploader, in console and also in seeds.rb file.

The gems I'm using for upload are:
 - carrierwave (2.2.2)
 - fog-aws (3.13.0)

What I have tried:
 - removing versions generation (nothing changed)
 - using RMagick instead of MiniMagick (nothing changed)
 - using storage :file (the upload was really fast)
 - installed Imagemagick version 6 (the latest is 7.1)
It seems the issue is related to the effective upload time, maybe the fog?

EDIT: 
after some analysis on logs I am pretty sure that the issue is due to the store time that is very long. 
Processing time is really fast, while cache time could be improved, but we may survive with it.

**Comment by courtsimas** [comment]
Is this a copy of https://github.com/carrierwaveuploader/carrierwave/issues/2615? do you need both open?

**Comment by robertoboldi** [comment]
sorry for the duplication, I got some issues of formatting of the question and I posted two of the same. Have you some ideas of the causes?

**Comment by robertoboldi** [comment]
Anyone has some ideas of the cause of this?
I also tried with no images attachments, and this issue continue to be for all kind of files.

**Comment by YSavir** [comment]
I've been experiencing this as well recently. Additionally any automated tests that interact with file uploads will fail because of timeouts, so this is more than an inconvenience but actually something impacting the development process.

By logging the various steps I could see that it happens during the `store` action. I haven't experimented further yet. I'll see what more I can discover, but if anyone else has any insights, they are very welcome.

**Comment by YSavir** [comment]
I solved the problem: all I had to do was post about it here. Just like that months worths of timeouts and sluggish uploads went away. No code changes, no dependency updates, nothing changed at all... but it's working smoothly now, at least in what I've tested so far.

No idea what happened there, but 🤷🏼. It's been a problem for the past few months and now it's fine. Wish I had further input. Only thing I can think of is that I added logging before and after each uploader hook and `puts`ed out the time.

**Comment by robertoboldi** [comment]
@YSavir , wow this may really change development times. 
Can you post your code so I try to test in my development env? Where exactly you put logging?
I really be thankful for it

**Comment by robertoboldi** [comment]
ok I have tested all the callbacks as you suggested, @YSavir. 
My store time is still really looooong, also the cache time is not really fast, but we can still survive with it. 

I added the logs at the remove hook and everything is still the same :-(

**Comment by patrickemuller** [comment]
I have been following the thread, but since I can't update our project to Rails 7 yet I can't contribute.
Considering it's only local, is there any chance that this is related to I/O in M1 Macs only?
Does anyone know if Rails 7 has changed (in any way) how it handles file saving?
@robertoboldi which version of Ruby are you using?

**Comment by emilebosch** [comment]
Yeah have a similar issue here. Its unbearable. I tried digging into the root cause with fog-aws but as soon as `directory.files` is called it just stalls forever. It was absolutely unworkable. Also the fog gem is hard to debug.I'm on an M1 on Rails 7.0.5.

Also  i was a big shocked about the huge amount of files and layers of indirecton needed to upload a file to s3 in aws/fog but that is maybe something for a different story.

**Comment by batalla3692** [comment]
I'm also on an M2 MacBook, and along with the slow store time, I was also getting `Excon::Error::Timeout` errors when uploading some really small files to S3 from my local. 

After reading [this issue](https://github.com/fog/fog-aws/issues/674) on `fog-aws` gem, which is kind of related, I figured out a way to bypass the timeouts by adding the following settings on the initializer of CarrierWave on my Rails project: 

```ruby
  CarrierWave.configure do |config|
    ...
    config.fog_credentials = {
      ...,
      connection_options: {
        write_timeout: 5
      }
    }
  end
```

Of course, this does not resolve the issue of the slow uploads, but at least I don't receive timeouts anymore, so I leave my solution here in case someone else has the same issue.

**Comment by robertoboldi** [comment]
Has anybody got any solution? 
I tried all Rails version until now (7.1.3)

this happens also when removing images, I don't know if this is clear.

**Comment by emilebosch** [comment]
@robertoboldi I just rolled my own upload and skippped all of this

**Comment by robertoboldi** [comment]
yes it seems a good idea. 
I may try do to this. Have you some good references to suggest?
I never done this, before

**Comment by emilebosch** [comment]
Something dumb like this would get u started

Some lib thing

```ruby
class Uploader
  def get_url(variant)
    "https://xxx.ams3.digitaloceanspaces.com/#{get_key(variant)}"
  end

  def initialize(model)
    @model = model
  end

  def magick(upload, &block)
    return unless upload
    @image = MiniMagick::Image.read(upload.tempfile)
    yield
    @image = nil
  end

  def variant(name, size)
    Tempfile.open { |xx|
      @image.resize(size)
      @image.write (xx)

      s3 = Aws::S3::Client.new(endpoint: "https://ams3.digitaloceanspaces.com")
      File.open(xx, "rb") do |file|
        s3.put_object(
          acl: "public-read",
          body: file,
          bucket: "xx",
          key: get_key(name),
        )
      end
    }
  end
end

class SpotUploader < Uploader
  def get_key(variant)
    "#{Rails.env}/spots/#{@model.id}/#{variant}.jpg"
  end

  def upload(upload)
    magick(upload) do
      variant(:thumb, "612x612")
      variant(:tiny, "50x50")
    end
  end
end
```

In the controller

```ruby
class PageController < Controller
  def create
    @spot = Spot.new(spot_params)
    @spot.user = User.first
    @spot.category = :building
    @spot.difficulty = "5"
    @spot.save

    loader = SpotUploader.new(@spot)
    loader.upload(spot_params[:file])
    @spot.file = loader.get_url(:thumb)
    @spot.save

    render json: merge_cover_photo(@spot)
  end
end
```
Gems

```ruby
gem "mini_magick"
gem "aws-sdk-s3"
```

**Comment by robertoboldi** [comment]
@emilebosch Wow, the full code!  
thank you so much and have a nice day

**Comment by emilebosch** [comment]
No worry remember this is crap and needs love and attention and mime types and security checks and all that jazz

**Comment by Xeej** [comment]
im trying change `X-Copy-From` to `Destination` with monkey patch, its work fasterer, but somtetimes to very long. on the graph we can see how much longer it took to load documents
<img width="333" alt="image" src="https://github.com/carrierwaveuploader/carrierwave/assets/55625846/b8e7f906-5b25-4cba-b662-349c2a859e56">

**Comment by Xeej** [comment]
Resolved for me by [site](https://atsss.medium.com/uploading-a-file-in-s3-via-carrierwave-creates-multiple-notification-events-e6ec1f85e3d4) If you see extra requests to cloud storage when loading, this should help you.
Just add in carrierwave config
```
CarrierWave.configure do |config|
  config.cache_storage = :file
end
```
I should read Realme carefully)

**Comment by robertoboldi** [comment]
thank you, it seems to really increase upload times, but I got strange behaviour: STILL ONLY IN LOCALHOST
- i have kept only the uploader (no image manipulation)
- se the cache_storage = :file in my carrierwave.rb initializer. 

I tried differeny image formats and dimensions from 100 KB to 1.5MB and literally got te time to upload and have the controller redirection. 
I got times from 2sec to 20sec and this is really SLOWER then before (uploads kept minimum 15-20sec - 10sec was very rare).  time was unlinked to file dimensions, I also tried always different files to avoid any type of caching. 

I'll read the readme as soon I have some time to dedicate to this.

...and thanks again Xeej!

**Comment by mshibuya** [comment]
Fixed by https://github.com/carrierwaveuploader/carrierwave/commit/6f5d6b80a99ab119000391a2cce3be76dbc7336f.
